src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt

```diff
@@ -158,296 +225,504 @@ class RsInlineFunctionProcessor(
-            blockParent.addRangeBefore(body.children.first(), body.children.last(), childContainingEnclosingStatement)
-            blockParent.addBefore(factory.createNewline(), childContainingEnclosingStatement)
-        } else {
-            blockParent.addStatements(body.children)
+        if (isUnitTestMode) error("Can't find containing statement")
+        return null
+    }
+
+    private fun PsiElement.wrapInBlockExprIfPossible(): RsStmt? {
+        val expr = when (val parent = parent) {
+            is RsMatchArm -> parent.expr
+            is RsLambdaExpr -> parent.expr
+            is RsConstant -> parent.expr
+            else -> return null
         }
+        if (expr != this) return null
+        (parent as? RsMatchArm)?.comma?.delete()
+        return wrapInBlockExpr()
     }
 
-    private fun replaceCallerWithRetExpr(body: RsBlock, caller: PsiElement) {
-        val retExpr = body.descendantsOfType<RsRetExpr>().firstOrNull()
-        if (retExpr != null) {
-            val retExprInside = retExpr.expr
-            if (retExprInside != null) {
+    private fun PsiElement.wrapInBlockExpr(): RsStmt? {
+        val blockExpr0 = RsPsiFactory(project).createBlockExpr("\n$text\n")
+        val blockExpr = replace(blockExpr0) as RsBlockExpr
+        return blockExpr.block.stmtList.singleOrNull()
+    }
+}
 
+private fun matchPatternsAndExpressions(patterns: List<RsPat>, expressions: List<RsExpr>): List<Pair<RsPat, RsExpr>> =
+    (patterns zip expressions).flatMap { (parameter, argument) ->
+        matchPatternAndExpression(parameter, argument)
+    }
 
-                // Covering a case in which ; isn't included in the expression, and parent is too wide.
-                // e.g. if RsExpr is surrounded by RsBlock going to the RsBlock won't help.
-                val actualRetExpr: PsiElement = if (retExpr.parent.text == retExpr.text + ";") {
-                    retExpr.parent
-                } else {
-                    retExpr
-                }
-                if (caller.parent !is RsBlock && caller.parent.text != caller.text + ";") {
-                    caller.replace(retExprInside)
-                    retExpr.delete()
-                } else {
-                    actualRetExpr.replace(retExprInside)
-                    if (caller.parent.text == caller.text + ";") {
-                        caller.parent.delete()
-                    } else {
-                        caller.delete()
-                    }
-                }
-            }
-        } else {
-            caller.delete()
+private fun matchPatternAndExpression(pattern: RsPat, expression: RsExpr): List<Pair<RsPat, RsExpr>> {
+    if (pattern is RsPatTup && expression is RsTupleExpr) {
+        val patList = pattern.patList
+        val exprList = expression.exprList
+        if (patList.size == exprList.size) {
+            return matchPatternsAndExpressions(patList, exprList)
         }
     }
+    // TODO: RsPatRef etc
```

Comment:
For those cases, it would also be nice to have "failing" tests if possible (or at least more detailed description in the comment)

Reply:
I actually forgot what does it mean, so removed